### PR TITLE
Add general support for PyTorch in Imperceptible ASR attack

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

This PR adds support to the framework-independent (PyTorch and TensorFlow) ASR attack.

Related to https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/273

**To-Do Checklist**:

* [ ] Test the attack with the PyTorch Deepspeech estimator
* [x] The `torch.stft` does not yet match the `tensorflow.signal.stft`. This needs to be identical so that both masking are identical
* [x] Move to `dev_1.5.0` base

## Type of change

Please check all the relevant options.

- [x] New feature (non-breaking)

# Testing

Attack comes with a set of tests.

**Test Configuration**:
- Fedora 31
- Python 3.8.5
- ART 1.5 development
- Torch 1.7.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 1]
```diff
@@ -58,11 +61,11 @@ class ImperceptibleASR(EvasionAttack):
         "batch_size",
     ]
 
-    _estimator_requirements = (TensorFlowV2Estimator, SpeechRecognizerMixin)
+    _estimator_requirements = (NeuralNetworkMixin, LossGradientsMixin, BaseEstimator, SpeechRecognizerMixin)
 
     def __init__(
         self,
-        estimator: "TensorFlowV2Estimator",
+        estimator: Union["PyTorchEstimator", "TensorFlowV2Estimator"],
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:68]: For typing we should create a new type in `art.utils` line 175 and call it "SPEECH_RECOGNIZER_TYPE" and make it a Union of `PyTorchDeepSpeech` and `TensorFlowLingvoAsr`.

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 2]
```diff
@@ -108,16 +107,34 @@ def __init__(
...
-        self._delta = tf1.placeholder(tf1.float32, shape=[None, None], name="art_delta")
-        self._power_spectral_density_maximum_tf = tf1.placeholder(tf1.float32, shape=[None, None], name="art_psd_max")
-        self._masking_threshold_tf = tf1.placeholder(
-            tf1.float32, shape=[None, None, None], name="art_masking_threshold"
-        )
-        # TensorFlow loss gradient ops
-        self._loss_gradient_masking_threshold_op_tf = self._loss_gradient_masking_threshold_tf(
-            self._delta, self._power_spectral_density_maximum_tf, self._masking_threshold_tf
-        )
+        if isinstance(self.estimator, TensorFlowV2Estimator):
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:110]: I think to make this check reliable for custom estimators should we search for `TensorFlowV2Estimator` in `self.estimator.__mro__, e.g. if a user inherits from `TensorFlowLingvoAsr`?

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 3]
```diff
@@ -108,16 +107,34 @@ def __init__(
...
+            )
+            self._masking_threshold_tf = tf1.placeholder(
+                tf1.float32, shape=[None, None, None], name="art_masking_threshold"
+            )
+            # TensorFlow loss gradient ops
+            self._loss_gradient_masking_threshold_op_tf = self._loss_gradient_masking_threshold_tf(
+                self._delta, self._power_spectral_density_maximum_tf, self._masking_threshold_tf
+            )
+
+        elif isinstance(self.estimator, PyTorchEstimator):
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:132]: Similar comment as for the `if` above.

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 4]
```diff
@@ -381,13 +441,33 @@ def _approximate_power_spectral_density_tf(
...
+        psd_matrix = gain_factor * torch.square(stft_matrix_abs / (2 * self._window_size))
 
-    def _loss_gradient_masking_threshold_torch(self):
-        """Compute loss gradient of the masking threshold loss in PyTorch."""
-        raise NotImplementedError
+        # approximate normalized psd: psd_matrix_approximated = 10^((96.0 - psd_matrix_max + psd_matrix)/10)
+        psd_matrix_approximated = pow(10.0, 9.6) / torch.unsqueeze(psd_maximum_stabilized, 1) * psd_matrix
+
+        # return PSD matrix such that shape is (batch_size, window_size // 2 + 1, frame_length)
+        # return torch.transpose(psd_matrix_approximated, 1, 2)
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:469]: Is this line code or comment?

Comment (hesseltuinhof) [reply art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:469]: Ups... Line 469 should have been removed :see_no_evil:

**Comment by hesseltuinhof** [comment]
Thank you @beat-buesser for the review! I have applied the review changes and rebased the PR onto `dev_1.5.0`.

**Comment by hesseltuinhof** [comment]
I have fixed the `torch.STFT` which was missing the correct `window` keyword. Now all STFT calls, whether `scipy.signal`, `tf.signal` or `torch`, return equal results, which is important to obtain the correct threshold masking loss :+1:

Furthermore, just while wrapping up PR #719 I introduced a bug by losing the line that updated the perturbation in stage 2. The current PR reverts the issue.

**Review by beat-buesser — APPROVED** [review]
